pkg/mutation/mutators/core/setter.go

```diff
@@ -0,0 +1,109 @@
+package core
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+)
+
+// Setter tells the mutate function what to do once we have found the
+// node that needs mutating.
+type Setter interface {
+	// SetValue takes the object that needs mutating and the key of the
+	// field on that object that should be mutated. It is up to the
+	// implementor to actually mutate the object.
+	SetValue(obj map[string]interface{}, key string) error
+
+	// KeyedListOkay returns whether this setter can handle keyed lists.
+	// If it can't, an attempt to mutate a keyed-list-type field will
+	// result in an error.
+	KeyedListOkay() bool
+
+	// KeyedListValue is the value that will be assigned to the
+	// targeted keyed list entry. Unlike SetValue(), this does
+	// not do mutation directly.
+	KeyedListValue() (map[string]interface{}, error)
+}
+
+var _ Setter = &defaultSetter{}
+
+func NewDefaultSetter(value interface{}) Setter {
+	return &defaultSetter{
+		value: value,
+	}
+}
+
+// defaultSetter is the default implementation of the Setter interface that supports
+// assigning plain values and external data placeholders.
+type defaultSetter struct {
+	value interface{}
+}
+
+func (s *defaultSetter) KeyedListOkay() bool { return true }
+
+func (s *defaultSetter) KeyedListValue() (map[string]interface{}, error) {
+	valueAsObject, ok := s.value.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("assign.value for keyed list is not an object: %+v", s.value)
+	}
+	return valueAsObject, nil
+}
+
+func (s *defaultSetter) SetValue(obj map[string]interface{}, key string) error {
+	value := s.value
+	incomingPlaceholder, isIncomingPlaceholder := value.(*types.ExternalDataPlaceholder)
+	if isIncomingPlaceholder {
+		// make a copy of the incoming placeholder so we can modify it
+		incomingPlaceholder = incomingPlaceholder.DeepCopy()
+	}
+
+	if _, ok := obj[key]; ok {
+		switch prev := obj[key].(type) {
+		case *types.ExternalDataPlaceholder:
+			if isIncomingPlaceholder {
+				incomingPlaceholder.ValueAtLocation = prev.ValueAtLocation
+			} else if valueAsString, err := encodeToString(value); err == nil {
+				// assign incoming value as string to the previous placeholder
+				incomingPlaceholder = prev.DeepCopy()
+				incomingPlaceholder.ValueAtLocation = valueAsString
+			} else {
+				// return an error if the incoming value can't be encoded to a string
+				return err
+			}
+		default:
+			if isIncomingPlaceholder {
+				valueAsString, err := encodeToString(prev)
+				if err != nil {
+					return err
+				}
+				incomingPlaceholder.ValueAtLocation = valueAsString
+			}
+		}
+	}
+
+	if incomingPlaceholder != nil {
+		value = incomingPlaceholder
+	}
+
+	obj[key] = value
+	return nil
+}
+
+func encodeToString(value interface{}) (string, error) {
```

Comment:
My thought was that since the value in `externaldata.Item` is `interface{}`, `"3" -> 3` is also considered idempotent but it makes sense that both key and value should have the same type to satisfy the requirement.

Reply:
Done. Since we might change the external data API in the future, I have versioned the `ExternalData` API by moving it from `pkg/mutation/types` to `apis/`. 

I also had to get rid of any reference of `pkg/mutation/types` in `apis/status/v1beta1` due to an import cycle caused by versioning the `ExternalData` API like I mentioned above.